### PR TITLE
dedupStoreHttpClient honors redirect timeout from client settings

### PR DESCRIPTION
**Pull request by kutsen99**
Read timeout from client settings
image

### CODE REVIEW

**Review by johnterickson — APPROVED** [review]
Double check to see if there are any other places we need to update

**Review by fadnavistanmay — APPROVED** [review]

**Comment by johnterickson** [comment]
>             var client = new DedupStoreClient(dedupStoreHttpClient, maxParallelism);

Needed here too?

---
Refers to: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:228 in 8ac4dfe. [](commit_id = 8ac4dfe10a306aa6d301ae78a41c18aeda7b35a3, deletion_comment = False)

File: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs [hunk 1]
```diff
@@ -124,7 +124,7 @@ private DedupManifestArtifactClientFactory()
                     domainId,
                     clientSettings,
                     context,
-                    cancellationToken);            
+                    cancellationToken);
         }
 
         public (DedupManifestArtifactClient client, BlobStoreClientTelemetry telemetry) CreateDedupManifestClient(
```
Comment (johnterickson) [inline src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:130]: >CreateDedupManifestClient [](http://example.com/codeflow?start=88&length=25)

Ideally `CreateDedupManifestClient` and `CreateDedupClientAsync` would both call a common function that creates and initializes an `IDedupStoreHttpClient`

Comment (kutsen99) [reply src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:130]: will be in another PR

**Comment by fadnavistanmay** [comment]
Adam, can you update the vss-api-netcore package version to be https://dev.azure.com/mseng/PipelineTools/_artifacts/feed/nugetvssprivate/NuGet/vss-api-netcore/overview/0.5.229-private
0.5.229-private as part of this PR please - https://github.com/microsoft/azure-pipelines-agent/blob/6e23e6bd044191997c5f512633e5f981166e01ed/src/Common.props#L13

**Comment by kutsen99** [comment]
> Adam, can you update the vss-api-netcore package version to be https://dev.azure.com/mseng/PipelineTools/_artifacts/feed/nugetvssprivate/NuGet/vss-api-netcore/overview/0.5.229-private 0.5.229-private as part of this PR please -
> 
> https://github.com/microsoft/azure-pipelines-agent/blob/6e23e6bd044191997c5f512633e5f981166e01ed/src/Common.props#L13

Yeah sure

File: src/Agent.Worker/Build/FileContainerServer.cs [hunk 2]
```diff
@@ -345,7 +345,13 @@ private async Task<UploadResult> BlobUploadAsync(IAsyncCommandContext context, I
...
                 int maxParallelism = context.GetHostContext().GetService<IConfigurationStore>().GetSettings().MaxDedupParallelism;
                 (dedupClient, clientTelemetry) = await DedupManifestArtifactClientFactory.Instance
-                    .CreateDedupClientAsync(verbose, (str) => context.Output(str), this._connection, maxParallelism, token);
+                    .CreateDedupClientAsync(
+                    verbose,
+                    (str) => context.Output(str),
+                    this._connection,
+                    maxParallelism,
+                    Microsoft.VisualStudio.Services.BlobStore.WebApi.Contracts.Client.BuildArtifact,
+                    token);
```
Comment (fadnavistanmay) [inline src/Agent.Worker/Build/FileContainerServer.cs:354]: is this the build artifact using blobstore as backend?

File: src/Microsoft.VisualStudio.Services.Agent/JobServer.cs [hunk 3]
```diff
@@ -173,7 +173,13 @@ public async Task<(DedupIdentifier dedupId, ulong length)> UploadAttachmentToBlo
         {
             int maxParallelism = HostContext.GetService<IConfigurationStore>().GetSettings().MaxDedupParallelism;
             var (dedupClient, clientTelemetry) = await DedupManifestArtifactClientFactory.Instance
-                    .CreateDedupClientAsync(verbose, (str) => Trace.Info(str), this._connection, maxParallelism, cancellationToken);
+                    .CreateDedupClientAsync(
+                        verbose,
+                        (str) => Trace.Info(str),
+                        this._connection,
+                        maxParallelism,
+                        null,
```
Comment (fadnavistanmay) [inline src/Microsoft.VisualStudio.Services.Agent/JobServer.cs:181]: nit: can you make this named parameter, agent:null, it isn't super clear why we are passing null here.

File: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs [hunk 4]
```diff
@@ -222,6 +237,8 @@ private DedupManifestArtifactClientFactory()
                 cancellationToken: cancellationToken,
                 continueOnCapturedContext: false);
 
+            dedupStoreHttpClient.SetRedirectTimeout(GetRedirectTimeoutFromClientSettings(clientSettings));
```
Comment (fadnavistanmay) [inline src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:240]: what happens if this value is null? Does it fallback to using env variable?

File: src/Agent.Worker/Build/FileContainerServer.cs (hunk 2 above)
Comment (kutsen99) [reply src/Agent.Worker/Build/FileContainerServer.cs:354]: yes

File: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs (hunk 4 above)
Comment (kutsen99) [reply src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:240]: env always comes first. if there's no env, null will reset it to the default

Comment (kutsen99) [reply src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:240]: https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=/BlobStore/Client/WebApi/DedupStoreHttpClient.cs&version=GBmaster&line=741&lineEnd=742&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

File: src/Microsoft.VisualStudio.Services.Agent/JobServer.cs (hunk 3 above)
Comment (kutsen99) [reply src/Microsoft.VisualStudio.Services.Agent/JobServer.cs]: done

File: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs (hunk 4 above)
Comment (fadnavistanmay) [reply src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:240]: Got it. Yeah, we will have to upgrade the vss-api-netcore pkg as well, in the next PR.

**Review by fadnavistanmay — APPROVED** [review]

File: src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs [hunk 5]
```diff
@@ -203,6 +209,15 @@ private DedupManifestArtifactClientFactory()
                 maxParallelism = DefaultDedupStoreClientMaxParallelism;
             }
             traceOutput($"Max dedup parallelism: {maxParallelism}");
+
+            ClientSettingsInfo clientSettings = clientType is null
+                ? null
```
Comment (johnterickson) [inline src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:214]: Is there not a client settings for default/unknown clienttype?

Comment (kutsen99) [reply src/Microsoft.VisualStudio.Services.Agent/Blob/DedupManifestArtifactClientFactory.cs:214]: there's no https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=/BlobStore/Client/WebApi/Contracts/ClientSettingsInfo.cs&version=GBmaster&line=22&lineEnd=23&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

**Pull request closed without merging**